lib/moonshot/command_line.rb

```diff
@@ -0,0 +1,108 @@
+require 'thor'
+
+module Moonshot
+  # This class implements the command-line `moonshot` tool.
+  class CommandLine
+    # TODO: No reason this is class-level, make this a for-real class that needs to be initialized.
+    def self.run!
+      # Find the Moonfile in this project.
+      orig_dir = Dir.pwd
+
+      loop do
+        break if File.exist?('Moonfile')
+
+        if Dir.pwd == '/'
+          warn 'Could not find Moonfile!'
+          exit 1
+        else
+          Dir.chdir('..')
+        end
+      end
+
+      moonfile_dir = Dir.pwd
+      Dir.chdir(orig_dir)
+
+      # Load any plugins and CLI extensions relative to the Moonfile
+      if File.directory?(File.join(moonfile_dir, 'moonshot'))
+        load_cli_extensions(moonfile_dir)
+        load_plugins(moonfile_dir)
+      end
+
+      Object.include(Moonshot::ArtifactRepository)
+      Object.include(Moonshot::BuildMechanism)
+      Object.include(Moonshot::DeploymentMechanism)
+      load(File.join(moonfile_dir, 'Moonfile'))
+
+      load_commands
+
+      # Determine what command is being run, which should be the first argument.
+      command = ARGV.shift
+      usage if %w[--help -h help].include?(command) or command.nil?
+
+      # Dispatch to that command, by executing it's parser, then
+      # comparing ARGV to the execute methods arity.
+      unless @commands.key?(command)
+        warn "Command not found: #{command}"
+        warn "\n"
+        usage
+      end
+
+      handler = @commands[command].new
+      handler.parser.parse!
+
+      unless ARGV.size == handler.method(:execute).arity
+        warn handler.parser.help
+        exit 1
+      end
+
+      handler.execute(*ARGV)
+    end
+
+    def self.register(klass)
+      @classes ||= []
+      @classes << klass
+    end
+
+    def self.registered_commands
+      @classes || []
+    end
+
+    def self.load_plugins(moonfile_dir)
+      plugins_path = File.join(moonfile_dir, 'moonshot', 'plugins', '**', '*.rb')
+      Dir.glob(plugins_path).each do |file|
+        load(file)
+      end
+    end
+
+    def self.load_cli_extensions(moonfile_dir)
+      cli_extensions_path = File.join(moonfile_dir, 'moonshot', 'cli_extensions', '**', '*.rb')
+      Dir.glob(cli_extensions_path).each do |file|
+        load(file)
+      end
+    end
+
+    def self.usage
+      warn "Usage: moonshot [command]"
+      warn
+      warn "Valid commands include:"
+      @commands.each do |c, k|
+        warn "#{c} #{k.description}"
+      end
+      exit 1
+    end
+
+    def self.load_commands
+      @commands = {}
+
+      # Include all Moonshot::Command and Moonshot::SSHCommand
+      # derived classes as subcommands, with the description of their
+      # default task.
+      registered_commands.each do |klass|
+        next unless klass.instance_methods.include?(:execute)
+
+        command_name = klass.to_s.split('::').last.scan(/[A-Z][a-z]+/).map(&:downcase).join('-')
```

Comment:
I point out active support because the common case everyone gets wrong is "Ruby-cased" acronyms.

```
=> "My::Project::::HTTPClient"
[4] pry(main)> class_name.split('::').last.scan(/[A-Z][a-z]+/).map(&:downcase).join('-')
=> "client"
```